components/patina_smbios/src/manager/protocol.rs

```diff
@@ -260,45 +313,72 @@ impl SmbiosProtocol {
 
         // SAFETY: Cast from protocol pointer to internal struct pointer
         let internal = unsafe { &*(protocol as *const SmbiosProtocolInternal) };
-        let manager = internal.manager.lock();
+
+        let result = {
+            let manager = internal.manager.lock();
+
+            // SAFETY: C UEFI protocol caller guarantees pointers are valid
+            unsafe {
+                let handle = *smbios_handle;
+                let type_filter = if record_type.is_null() { None } else { Some(*record_type) };
+
+                // Use the iterator to find the next record
+                let mut iter = manager.iter(type_filter);
+
+                // Skip records until we find the one after the current handle
+                let next_record = if handle == SMBIOS_HANDLE_PI_RESERVED {
+                    // Starting iteration - get first record
+                    iter.next()
+                } else {
+                    // Find the record after the current handle
+                    iter.skip_while(|(hdr, _)| hdr.handle != handle).nth(1)
+                };
+
+                match next_record {
+                    Some((header_value, prod_handle)) => {
+                        let found_handle = header_value.handle;
+
+                        // Return tuple with all the data we need
+                        (found_handle, header_value, prod_handle, true)
+                    }
+                    None => (SMBIOS_HANDLE_PI_RESERVED, SmbiosTableHeader::new(0, 0, 0), None, false),
+                }
+            }
+        }; // manager borrow drops here
 
         // SAFETY: The pointers are checked for being null above. They are expected to point
         // to valid data objects due to the expectations of the protocol interface. The type_filter
         // is optionally dereferenced after a null check.
         unsafe {
```

Comment:
Consider narrowing the scope of this unsafe as well so it covers as little code as possible.